src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs

```diff
@@ -92,6 +92,23 @@ internal sealed class MsQuicContextSafeHandle : MsQuicSafeHandle
     /// </summary>
     private readonly MsQuicSafeHandle? _parent;
 
+#if DEBUG
+    /// <summary>
+    /// Native memory to hold TLS secrets. It needs to live same cycle as the underlying connection.
+    /// </summary>
+    private unsafe QUIC_TLS_SECRETS* _tlsSecrets;
```

Comment:
`MsQuicContextSafeHandle` is common for stream, connection and listener. Logically, this pointer doesn't belong here as it is specific for a connection.

If you need to de-allocate with connection's `MsQuicApi.Api.ApiTable->ConnectionClose` you could make `MsQuicTlsSecret` a `SafeHandle` and make it parent of connection's `MsQuicContextSafeHandle`. That will increment/decrement ref count with connection handle lifetime and ensure that `MsQuicTlsSecret` outlives it (same way as connection is stream's parent). This would require type change for  the `_parent` field: `MsQuicSafeHandle? parent --> SafeHandle? parent`, but it should be doable as we're not using anything specific for `MsQuicSafeHandle`.

BTW, could we file this as an MsQuic issue? Are there any plans to change the lifetime requirements or ownership of this object? Or is this by design and expected for C/C++ API?

Reply:
I was thinking about new SafeHandle but I decided not to to keep it simple. While logically I agree with you I don't think it matters for Debug builds. e.g. the downside is extra unused filed for objects that do not need that. If you want this really clean we can derive separate SafeHandle for the connection.

As far as the MsQuic behavior I have no strong feelings. I think existing behavior would probably work for most cases. We _could_ have single static buffer as placeholder for any lingering operations. But that may leave secrets around - again not sure if that is big deal for Debug builds as the whole secret dumping is insecure. 

cc: @nibanks for more thoughts.

Reply:
So I thought about it, and I think I don't mind an extra pointer being in base safe handle given it's debug-only... But I was thinking it could be named as something more basic, like `DebugBuffer`, so `public unsafe QUIC_TLS_SECRETS* GetSecretsBuffer()` would become `public unsafe void* GetDebugBuffer(int byteCount)` and `MsQuicTlsSecrets` class would call `QUIC_TLS_SECRETS* ptr = (QUIC_TLS_SECRETS*)handle.GetDebugBuffer(sizeof(QUIC_TLS_SECRETS));` -- WDYT? It might be once again a bit of overengineering, but it removes logical "strain" of TlsSecrets being present in base class

Reply:
I left the name unchanged - to the @ManickaP  point that it describes exactly what it is.